GANDLF/models/lightning_module.py

```diff
@@ -42,4 +138,381 @@ def configure_optimizers(self):
+
+        return images, labels
+
+    def _handle_dynamic_batch_size_in_differential_privacy_mode(self, subject):
+        raise NotImplementedError(
+            "Differential privacy is not implemented yet in lightning version"
+        )
+
+    def _initialize_differential_privacy(self):
+        raise NotImplementedError(
+            "Differential privacy is not implemented yet in lightning version"
+        )
+
+    def on_train_epoch_start(self):
+        self._set_epoch_start_time()
+        if self.params["track_memory_usage"]:
+            self._write_epoch_start_process_resource_usage(self.current_epoch)
+        if self.params["verbose"]:
+            self._print_epoch_start_time()
+
+    def _write_epoch_start_process_resource_usage(self, epoch):
+        filename = f"memory_usage_local_rank_{self.local_rank}_global_rank_{self.global_rank}.csv"
+        memory_stats_dir = self._prepare_memory_stats_save_dir()
+        full_filepath = os.path.join(memory_stats_dir, filename)
+        file_write_mode = "a" if os.path.exists(full_filepath) else "w"
+        using_cuda = "cuda" in self.device.type
+
+        memory_info_string = "Epoch,Memory_Total,Memory_Available,Memory_Percent_Free,Memory_Usage,"  # used to write output
+        if using_cuda:
+            memory_info_string += (
+                "CUDA_active.all.peak,CUDA_active.all.current,CUDA_active.all.allocated"
+            )
+        memory_info_string += "\n"
+
+        host_memory_stats = psutil.virtual_memory()
+        memory_info_string += (
+            str(epoch)
+            + ","
+            + str(host_memory_stats[0])
+            + ","
+            + str(host_memory_stats[1])
+            + ","
+            + str(host_memory_stats[2])
+            + ","
+            + str(host_memory_stats[3])
+        )
+        if using_cuda:
+            cuda_memory_stats = torch.cuda.memory_stats()
+            memory_info_string += (
+                ","
+                + str(cuda_memory_stats["active.all.peak"])
+                + ","
+                + str(cuda_memory_stats["active.all.current"])
+                + ","
+                + str(cuda_memory_stats["active.all.allocated"])
+            )
+        memory_info_string += ",\n"
+        with open(full_filepath, file_write_mode) as file_mem:
+            file_mem.write(memory_info_string)
+
+    @rank_zero_only
+    def _prepare_memory_stats_save_dir(self):
+        memory_stats_dir = os.path.join(self.output_dir, "memory_stats")
+        os.makedirs(memory_stats_dir, exist_ok=True)
+        return memory_stats_dir
+
+    @rank_zero_only
+    def _print_epoch_start_time(self):
+        print("Epoch start time : ", get_date_time(), flush=True)
+
+    @rank_zero_only
+    def _set_epoch_start_time(self):
+        self.epoch_start_time = time.time()
+
+    # TODO when used with multple GPUs, thil will produce multiple logs
+    # for each GPU. We should think on doing allgather here in a function
```

Comment:
```suggestion
    # for each GPU. We should think on doing altogether here in a function
```

Reply:
I actually really meant allgather operation for retrieving everything from metricas

Reply:
Ah okay. Then maybe add that to the spell checker?

Reply:
Done